File: src/ast/analyzing/langs/astml/fact.ml
(*
   Copyright 2012-2020 Codinuum Software Lab <https://codinuum.com>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)
(* fact extractor *)



module A = Astml

module CCX (L : Label.T) = struct

  module FB = Fact_base.F(L)
  open FB

  let getlab = getlab

  let p_in_decl    = mkccxres "inDeclaration"
  let p_in_fun_def = mkccxres "inFunctionDefinition"
  let p_in_stmt    = mkccxres "inStatement"
  let p_in_if_stmt = mkccxres "inIfStatement"

  let p_name       = mkccxres "name"
  let p_value      = mkccxres "value"
  let p_operator   = mkccxres "operator"

  let node_filter options nd =
    if options#fact_restricted_flag then
(*      let lab = getlab nd in *)
      true
    else
      true

  let node_pair_filter options nd1 nd2 =
    (node_filter options nd1) && (node_filter options nd2)


  class extractor options cache_path tree = object (self)
    inherit extractor_base options cache_path tree as super

    method id = "Astml"

    method scanner_body_after_subscan nd lab entity =

      if node_filter options nd then begin

	self#add (entity, p_is_a, mkccxres nd#data#get_category);

        if L.CCX.is_decl lab || L.CCX.is_translation_unit lab then begin
          self#add (entity, p_in_file, self#fileentity);
          self#set_version entity
        end;

        begin
          try
            self#add (entity, p_name, mklit (L.get_name lab))
          with
            Not_found -> ()
        end;

        begin
          try
            self#add (entity, p_value, mklit (L.get_value lab))
          with
            Not_found -> ()
        end;

        begin
          try
            self#add (entity, p_operator, mklit (L.get_operator lab))
          with
            Not_found -> ()
        end;

        self#add_surrounding_xxx L.CCX.is_decl nd entity p_in_decl;
        self#add_surrounding_xxx L.CCX.is_fun_def nd entity p_in_fun_def;
        self#add_surrounding_xxx L.CCX.is_stmt nd entity p_in_stmt;

      end

  end (* of class Astml.Fact.CCX.extractor *)

  (* main function *)
  let extract options cache_path tree =
    try
      let extractor = new extractor options cache_path tree in
      DEBUG_MSG "setting lang prefix to %s" Astml.ccx_prefix;
      extractor#set_lang_prefix Astml.ccx_prefix;
      DEBUG_MSG "extracting...";
      extractor#extract
    with
    | Triple.File_exists s -> Common.warning_msg "file exists: \"%s\"" s
    | Triple.Lock_failed -> Common.warning_msg "fact buffer is already locked."


end (* of functor Astml.Fact.CCX *)

File: python/src/cca/factutil/fact.py
#!/usr/bin/env python3

'''
  Factutil: helper scripts for source code entities

  Copyright 2012-2021 Codinuum Software Lab <https://codinuum.com>

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
'''

import logging

from .const import RELEASE_NS, SVNREV_NS, GITREV_NS, GUARD_NS
from .rdf import Graph, Resource, Predicate


logger = logging.getLogger()


class Fact(Graph):

    def create_release_version(self, rel):
        return Resource(RELEASE_NS + rel)

    def create_svn_revision(self, rev):
        s = '%s%s' % (SVNREV_NS, rev)
        return Resource(s)

    def create_git_revision(self, rev):
        s = '%s%s' % (GITREV_NS, rev)
        return Resource(s)

    def get_guard_pred(self, pred):
        g_pred = None
        try:
            g_pred = self._g_pred_map[pred]

        except KeyError:
            g_pred = Predicate(GUARD_NS+'?orig='+pred.get_namespace(),
                               pred.get_local_name())
            self._g_pred_map[pred] = g_pred

        return g_pred

    def list_guards(self, s, p, o):
        guards = []
        gp = self.get_guard_pred(p)
        q = self._create_statement(None, gp, s)
        for stmt in self.find_statements(q):
            g = Resource(node=stmt.subject)
            if self.contains(g, gp, o):
                guards.append(g)
        return guards

    def add(self, subj, pred, obj, attr=None, value=None):
        self._add(subj, pred, obj)
        if attr and value:
            blk = None
            guards = self.list_guards(subj, pred, obj)

            if len(guards) == 0:
                blk = Resource()
                g_pred = self.get_guard_pred(pred)
                self._add(blk, g_pred, subj)
                self._add(blk, g_pred, obj)

            else:
                blk = guards[0]

            if blk is None:
                blk = Resource()

            self._add(blk, attr, value)

    def addStatement(self, stmt, attr=None, value=None):
        self.add(stmt.subject, stmt.predicate, stmt.object, attr, value)
